torch/quantization/fake_quantize.py

```diff
@@ -79,7 +79,12 @@ def forward(self, X):
         if self.observer_enabled:
             self.activation_post_process(X.detach())
             _scale, _zero_point = self.calculate_qparams()
-            self.scale, self.zero_point = _scale.to(self.scale.device), _zero_point.to(self.zero_point.device)
+            _scale, _zero_point = _scale.to(self.scale.device), _zero_point.to(self.zero_point.device)
+            self.scale.resize_(_scale.shape)
+            self.scale.copy_(_scale)
```

Comment:
Are these `copy_` operations dangerous in case the parameters are updated simultaneously over multiple GPUs (say two GPUs try to update "at the same time")?

Reply:
Example showcasing dubious behavior

```python
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0, 1'

import torch
import torch.nn as nn

class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.fc1 = nn.Linear(4, 3)
        self.fc2 = nn.Linear(3, 2)

    def forward(self, x):
        self.fc1.myattribute.copy_(self.net.fc1.myattribute + 3)
        print(self.fc1.myattribute)
        x = self.fc1(x)
        x = self.fc2(x)
        return x

globnet = M()

globnet.fc1.myattribute = torch.tensor(0)

input = torch.randn(4, 4, 4, 4)

device_ids = [0, 1]

data_scattered = nn.parallel.scatter(input, device_ids, dim=0)
# transfer the parameters to all GPUs
replicas = nn.parallel.replicate(globnet.cuda(), device_ids)

outputs = nn.parallel.parallel_apply(replicas, data_scattered)
gathered = nn.parallel.gather(outputs, 0)

print('end', globnet.net.fc1.myattribute)
```

This prints at times
```
tensor(3)
tensor(3)
end tensor(3)
```

and sometimes

```
tensor(6)
tensor(6)
end tensor(6)
```

What do you think?

Reply:
ping @vkuzo